cmd/clusterctl/pkg/client/cluster/inventory_managementgroup.go

```diff
@@ -17,19 +17,50 @@ limitations under the License.
 package cluster
 
 import (
-	"fmt"
 	"strings"
 
 	"github.com/pkg/errors"
 	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
 )
 
-func (p *inventoryClient) GetManagementGroups() ([]ManagementGroup, error) {
-	providerList, err := p.list()
-	if err != nil {
-		return nil, err
+// ManagementGroup is a group of providers composed by a CoreProvider and a set of Bootstrap/ControlPlane/Infrastructure providers
+// watching objects in the same namespace. For example, a management group can be used for upgrades, in order to ensure all the providers
+// in a management group support the same ClusterAPI version.
+type ManagementGroup struct {
+	CoreProvider clusterctlv1.Provider
+	Providers    []clusterctlv1.Provider
+}
+
+// Equals return true if two management groups have the same core provider.
+func (mg *ManagementGroup) Equals(other *ManagementGroup) bool {
+	return mg.CoreProvider.Equals(other.CoreProvider)
+}
+
+// GetProviderByInstanceName returns a specific provider instance.
+func (mg *ManagementGroup) GetProviderByInstanceName(instanceName string) *clusterctlv1.Provider {
+	for _, provider := range mg.Providers {
+		if provider.InstanceName() == instanceName {
+			return &provider
+		}
 	}
+	return nil
+}
 
+// ManagementGroupList defines a list of management groups
+type ManagementGroupList []ManagementGroup
+
+// FindManagementGroupByProviderInstanceName return the management group that hosts a given provider.
+func (ml *ManagementGroupList) FindManagementGroupByProviderInstanceName(instanceName string) *ManagementGroup {
+	for _, managementGroup := range *ml {
+		if p := managementGroup.GetProviderByInstanceName(instanceName); p != nil {
+			return &managementGroup
+		}
+	}
+	return nil
+}
+
+// deriveManagementGroups derives the management groups from a list of providers.
+func deriveManagementGroups(providerList *clusterctlv1.ProviderList) (ManagementGroupList, error) {
```

Comment:
splitting `deriveManagementGroups ` from reading providers so I can use to validate providers contracts during init